datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaBasedConnectorTaskMetrics.java

```diff
@@ -54,28 +55,48 @@
   private static final Map<String, AtomicLong> AGGREGATED_NUM_AUTO_PAUSED_PARTITIONS_WAITING_FOR_DEST_TOPIC =
       new ConcurrentHashMap<>();
 
-  private final AtomicLong _numConfigPausedPartitions = new AtomicLong(0);
-  private final AtomicLong _numAutoPausedPartitionsOnError = new AtomicLong(0);
-  private final AtomicLong _numAutoPausedPartitionsOnInFlightMessages = new AtomicLong(0);
-  private final AtomicLong _numAutoPausedPartitionsAwaitingDestTopic = new AtomicLong(0);
-  private final AtomicLong _numTopics = new AtomicLong(0);
+  private static final Map<String, AtomicLong> NUM_TOPICS_PER_METRIC_KEY = new ConcurrentHashMap<>();
+  private static final Map<String, AtomicLong> NUM_CONFIG_PAUSED_PARTITIONS_PER_METRIC_KEY = new ConcurrentHashMap<>();
+  private static final Map<String, AtomicLong> NUM_AUTO_PAUSED_PARTITIONS_ON_ERROR_PER_METRIC_KEY =
+      new ConcurrentHashMap<>();
+  private static final Map<String, AtomicLong> NUM_AUTO_PAUSED_PARTITIONS_ON_INFLIGHT_MESSAGES_PER_METRIC_KEY =
+      new ConcurrentHashMap<>();
+  private static final Map<String, AtomicLong> NUM_AUTO_PAUSED_PARTITIONS_WAITING_FOR_DEST_TOPIC_PER_METRIC_KEY =
+      new ConcurrentHashMap<>();
+
+  private long _numConfigPausedPartitions = 0;
+  private long _numAutoPausedPartitionsOnError = 0;
+  private long _numAutoPausedPartitionsOnInFlightMessages = 0;
+  private long _numAutoPausedPartitionsAwaitingDestTopic = 0;
+  private long _numTopics = 0;
```

Comment:
Same concern about these no longer being atomic, and what happens if multiple threads try to update any of these at the same time? Can't they race and cause incorrect updates?

Reply:
changed it to AtomicLong.